docs/pages/style-spec.js

```diff
@@ -1275,6 +1276,57 @@ export default class extends React.Component {
                                 and a camera expression, and must adhere to the restrictions described above for both.</p>
 
+                            <h3>Type system</h3>
+                            <p>The input arguments to expressions, and their result values, use the same set of <a
+                                href="#types">types</a> as the rest of the style specification: boolean, string,
+                                number, color, and arrays of these types. Furthermore, expressions are <em>type safe</em>:
+                                each use of an expression has a known result type and required argument types, and the
+                                SDKs verify that the result type of an expression is appropriate for the context in
+                                which it is used. For example, the result type of an expression in the <a
+                                href="#layer-filter"><code>filter</code></a> property must be <a
+                                href="#types-boolean">boolean</a>, and the arguments to the <a
+                                href="#expressions-+"><code>+</code></a> operator must be <a
+                                href="#types-number">numbers</a>.
+                            </p>
+
+                            <p>When working with feature data, the type of a feature property value is typically not known
+                                ahead of time by the SDK. In order to preserve type safety, when evaluating a data
+                                expression, the SDK will check that the property value is appropriate for the context.
+                                For example, if you use the expression <code>["get", "feature-color"]</code> for the <a
+                                    href="#paint-circle-circle-color"><code>circle-color</code></a> property, the SDK
+                                will verify that the <code>feature-color</code> value of each feature is a string
+                                identifying a valid <a href="#types-color">color</a>. If this check fails, an error will
+                                be indicated in an SDK-specific way (typically a log message), and the default value for
+                                the property will be used instead.
+                            </p>
+
+                            <p>In most cases, this verification will occur automatically wherever it is needed. However,
+                                in certain situations, the SDK may be unable to automatically determine the expected
+                                result type of a data expression from surrounding context. For example, it is not clear
+                                whether the expression <code>["&lt;", ["get", "a"], ["get", "b"]]</code> is attempting
+                                to compare strings or numbers. In situations like this, you must use one of
```

Comment:
"must" could be misleading, since using a conversion rather than an assertion is also an option.

Reply:
How about:

> In situations like this, the simplest option is to use <em>type assertions</em>: <code>["&lt;", ["number", ["get", "a"]], ["number", ["get", "b"]]]</code>.  A type assertion <em>checks</em> that the feature data actually matches the expected type of the data expression.  If this check fails, it produces an error and causes the whole expression to fall back to the default value for the property being defined.  The assertion operators are ...
>
> Another option is to use a <em>type conversion</em>.  Expressions only perform one kind of implicit type conversion...

This repeats the description of runtime type checking from the previous paragraph, but I think the redundancy is worthwhile in order to explicitly state what an assertion does.

Reply:
Hmm, not convinced about this change. I think we should keep the use cases clearly separated here: conversion is for when you have one type, but need another. Assertion is for when the implementation needs you to tell it what type you have. If you are using a conversion, then you're not in a situation where the implementation is unable to automatically determine the type.

Reply:
> If you are using a conversion, then you're not in a situation where the implementation is unable to automatically determine the type.

You could be in such a situation -- e.g., say you want to do `["==", ["id"], ["get", "name"]]`.  Depending on your data, you really might want to use conversion rather than assertion as your way out of the type checking error.

That said, I'm 👍 for focusing on assertions as the primary way to deal with this, since, though conversions _can_ be used in this way, their typical use cases are much broader.    I still think the word "must" is misleading, and also that we should include an explicit description of assertions' behavior.